### PR TITLE
Remove the copy-procedure.

### PR DESCRIPTION
**Pull request by kffaxen**
This patch replaces the 'copy' procedure with assignments or a
'copy' function, but inly for arrays.

### CODE REVIEW

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 1]
```diff
@@ -50,62 +50,70 @@ type Location = Maybe (Expression ())
...
-  , l@(ConstExpr (IntConst n _)) <- arrayLength arg2
-  = if n < safetyLimit opts
-      then initArray (Just arg1) l:map (\i -> Assign (ArrayElem arg1 [litI32 i]) (ArrayElem arg2 [litI32 i])) [0..(n-1)]
-      else error $ unlines ["Internal compiler error: array size (" ++ show n ++ ") too large for deepcopy", show arg1, show arg2]
-
   | StructType _ fts <- typeof arg2
   = concatMap (deepCopyField . fst) fts
 
-  | not (isArray (typeof arg1))
+  | not (isArray (typeof arg1) || isNativeArray (typeof arg1))
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: Missing not before isNativeArray?

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 2]
```diff
@@ -50,62 +50,70 @@ type Location = Maybe (Expression ())
...
+flattenCopy dst (t:ts) (l:ls) cLen = call "copyArrayPos" [ValueParameter dst, ValueParameter cLen, ValueParameter t]
                                    : flattenCopy dst ts ls (ePlus cLen l)
 
 ePlus :: Expression () -> Expression () -> Expression ()
 ePlus (ConstExpr (IntConst 0 _)) e = e
 ePlus e (ConstExpr (IntConst 0 _)) = e
 ePlus e1 e2                        = binop (1 :# (NumType Signed S32)) "+" e1 e2
 
+-- | Lower array copy
+
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: Skip blank line.

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 3]
```diff
@@ -50,62 +50,70 @@ type Location = Maybe (Expression ())
...
   case inExp of
-    [x] | outExp == x -> Empty
-    _                 -> call "copy" (map ValueParameter (outExp:inExp))
+    [x] | outExp == x -> []
+    _                 -> deepCopy opts outExp inExp
 
 -- | Expand copying of aggregate data to copying of components
-deepCopy :: Options -> [ActualParameter ()] -> [Program ()]
-deepCopy opts [ValueParameter arg1, ValueParameter arg2]
+deepCopy :: Options -> Expression () -> [Expression ()] -> [Program ()]
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: Can Options be removed as parameter to this function and all the "ask backendOpts" in compileProg disappear?

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 1 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: No, I think the 'not' is where it should be, outside the parenthesis.

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 2 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: Ok, will do.

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 3 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: Yes, I missed that simplification. It is only needed when lowering. Will do!

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 4]
```diff
@@ -56,56 +56,63 @@ copyProg Nothing _ = Empty
...
+
+  | otherwise
+  = [ initArray (Just arg1) expDstLen, copyFirstSegment ] ++
+      flattenCopy arg1 ins argnLens arg1len
+    where expDstLen = foldr ePlus (litI32 0) aLens
+          copyFirstSegment = if arg1 == in1
+                                then Empty
+                                else call "copyArray" [ ValueParameter arg1
+                                                      , ValueParameter in1]
+          aLens@(arg1len:argnLens) = map (\src -> arrayLength src) ins'
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: map arrayLength ins'?

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 5]
```diff
@@ -56,56 +56,63 @@ copyProg Nothing _ = Empty
...
+lowerCopy opts NativeArray{} arg1 [arg2]
+  | l@(ConstExpr (IntConst n _)) <- arrayLength arg2
+  = if n < safetyLimit opts
+      then initArray (Just arg1) l:map (\i -> Assign (ArrayElem arg1 [litI32 i]) (ArrayElem arg2 [litI32 i])) [0..(n-1)]
+      else error $ unlines ["Internal compiler error: array size (" ++ show n ++ ") too large for deepcopy", show arg1, show arg2]
+lowerCopy opts t e es = error $ "Frontend.lowerCopy: funny type (" ++ show t ++ ") or destination\n"
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: Use "_" as name for opts and other unused parameters in the LHS.

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 1 above)
Comment (pjonsson) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: Yeah, I'm blind.

File: src/Feldspar/Compiler/Imperative/Frontend.hs [hunk 6]
```diff
@@ -56,56 +56,63 @@ copyProg Nothing _ = Empty
...
+          copyFirstSegment = if arg1 == in1
+                                then Empty
+                                else call "copyArray" [ ValueParameter arg1
+                                                      , ValueParameter in1]
+          aLens@(arg1len:argnLens) = map (\src -> arrayLength src) ins'
+lowerCopy opts NativeArray{} arg1 [arg2]
+  | l@(ConstExpr (IntConst n _)) <- arrayLength arg2
+  = if n < safetyLimit opts
+      then initArray (Just arg1) l:map (\i -> Assign (ArrayElem arg1 [litI32 i]) (ArrayElem arg2 [litI32 i])) [0..(n-1)]
+      else error $ unlines ["Internal compiler error: array size (" ++ show n ++ ") too large for deepcopy", show arg1, show arg2]
```
Comment (pjonsson) [inline src/Feldspar/Compiler/Imperative/Frontend.hs]: Use "lowerCopy" to refer to the function failing in the error message.

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 4 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: Absolutely!

File: src/Feldspar/Compiler/Imperative/Frontend.hs (hunk 6 above)
Comment (kffaxen) [reply src/Feldspar/Compiler/Imperative/Frontend.hs]: Yup.

**Pull request merged**